staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/managedfieldsupdater_test.go

```diff
@@ -29,42 +30,255 @@ import (
 		"apiVersion": "v1",
-		"kind": "Pod",
+		"kind": "ConfigMap",
 		"metadata": {
-			"name": "pod",
-			"labels": {"app": "nginx2"}
+			"name": "configmap"
 		},
-	}`), &obj2.Object); err != nil {
-		t.Fatalf("error decoding YAML: %v", err)
+		"data": {
+			"key": "value"
+		}
+	}`))
+	previousManagedFields := f.ManagedFields()
+
+	time.Sleep(time.Second)
+
+	updateObject(t, &f, "fieldmanager_test", []byte(`{
+		"apiVersion": "v1",
+		"kind": "ConfigMap",
+		"metadata": {
+			"name": "configmap"
+		},
+		"data": {
+			"key": "value"
+		}
+	}`))
+	newManagedFields := f.ManagedFields()
+
+	if !reflect.DeepEqual(previousManagedFields, newManagedFields) {
+		t.Errorf("ManagedFields changed:\nBefore:\n%v\nAfter:\n%v", previousManagedFields, newManagedFields)
 	}
+}
+
+func TestManagedFieldsApplyWithoutChangesDoesNotModifyTime(t *testing.T) {
+	f := NewTestFieldManager(schema.FromAPIVersionAndKind("v1", "ConfigMap"), "", nil)
+
+	applyObject(t, &f, "fieldmanager_test", []byte(`{
+		"apiVersion": "v1",
+		"kind": "ConfigMap",
+		"metadata": {
+			"name": "configmap"
+		},
+		"data": {
+			"key": "value"
+		}
+	}`))
+	previousManagedFields := f.ManagedFields()
+
 	time.Sleep(time.Second)
-	if err := f.Update(obj2, "fieldmanager_test"); err != nil {
-		t.Fatalf("failed to update object: %v", err)
+
+	applyObject(t, &f, "fieldmanager_test", []byte(`{
+		"apiVersion": "v1",
+		"kind": "ConfigMap",
+		"metadata": {
+			"name": "configmap"
+		},
+		"data": {
+			"key": "value"
+		}
+	}`))
+	newManagedFields := f.ManagedFields()
+
+	if !reflect.DeepEqual(previousManagedFields, newManagedFields) {
+		t.Errorf("ManagedFields changed:\nBefore:\n%v\nAfter:\n%v", previousManagedFields, newManagedFields)
+	}
+}
+
+func TestNonManagedFieldsUpdateDoesNotModifyTime(t *testing.T) {
+	f := NewTestFieldManager(schema.FromAPIVersionAndKind("v1", "ConfigMap"), "", nil)
+
+	updateObject(t, &f, "fieldmanager_a_test", []byte(`{
+		"apiVersion": "v1",
+		"kind": "ConfigMap",
+		"metadata": {
+			"name": "configmap"
+		},
+		"data": {
+			"key_a": "value"
+		}
+	}`))
+	updateObject(t, &f, "fieldmanager_b_test", []byte(`{
+		"apiVersion": "v1",
+		"kind": "ConfigMap",
+		"metadata": {
+			"name": "configmap"
+		},
+		"data": {
+			"key_b": "value"
+		}
+	}`))
+	previousManagedFields := f.ManagedFields()
+	previousEntries := map[string]v1.ManagedFieldsEntry{}
+	for _, entry := range previousManagedFields {
+		previousEntries[entry.Manager] = entry
+	}
+
+	time.Sleep(time.Second)
+
+	updateObject(t, &f, "fieldmanager_a_test", []byte(`{
+		"apiVersion": "v1",
+		"kind": "ConfigMap",
+		"metadata": {
+			"name": "configmap"
+		},
+		"data": {
+			"key_a": "value",
+			"key_b": "new-value"
+		}
+	}`))
+	newManagedFields := f.ManagedFields()
+	newEntries := map[string]v1.ManagedFieldsEntry{}
+	for _, entry := range newManagedFields {
+		newEntries[entry.Manager] = entry
+	}
+
+	if reflect.DeepEqual(previousManagedFields, newManagedFields) {
```

Comment:
This test isn't really needed since it'll fail below anyway, right? (same thing applies to the next test for apply)